pkg/controllers/management/auth/project_cluster/cluster_handler.go

```diff
@@ -0,0 +1,417 @@
+		}
+	}
+
+	d, err := json.Marshal(annoMap)
+	if err != nil {
+		return obj, err
+	}
+
+	// Save the required role templates to the annotation on the obj
+	if meta.GetAnnotations() == nil {
+		meta.SetAnnotations(make(map[string]string))
+	}
+	meta.GetAnnotations()[roleTemplatesRequired] = string(d)
+	return obj, nil
+}
+
+func (l *clusterLifecycle) reconcileClusterCreatorRTB(obj runtime.Object) (runtime.Object, error) {
+	return apisv3.CreatorMadeOwner.DoUntilTrue(obj, func() (runtime.Object, error) {
+		metaAccessor, err := meta.Accessor(obj)
+		if err != nil {
+			return obj, err
+		}
+
+		typeAccessor, err := meta.TypeAccessor(obj)
+		if err != nil {
+			return obj, err
+		}
+
+		creatorID, ok := metaAccessor.GetAnnotations()[CreatorIDAnn]
+		if !ok || creatorID == "" {
+			logrus.Warnf("%v %v has no creatorId annotation. Cannot add creator as owner", typeAccessor.GetKind(), metaAccessor.GetName())
+			return obj, nil
+		}
+		cluster := obj.(*apisv3.Cluster)
+
+		if apisv3.ClusterConditionInitialRolesPopulated.IsTrue(cluster) {
+			// The clusterRoleBindings are already completed, no need to check
+			return obj, nil
+		}
+
+		roleJSON, ok := cluster.Annotations[roleTemplatesRequired]
+		if !ok {
+			return cluster, nil
+		}
+
+		roleMap := make(map[string][]string)
+		err = json.Unmarshal([]byte(roleJSON), &roleMap)
+		if err != nil {
+			return obj, err
+		}
+
+		var createdRoles []string
+
+		for _, role := range roleMap["required"] {
+			rtbName := "creator-" + role
+
+			if rtb, _ := l.crtbLister.Get(metaAccessor.GetName(), rtbName); rtb != nil {
+				createdRoles = append(createdRoles, role)
+				// This clusterRoleBinding exists, need to check all of them so keep going
+				continue
+			}
+
+			// The clusterRoleBinding doesn't exist yet so create it
+			om := v1.ObjectMeta{
+				Name:      rtbName,
+				Namespace: metaAccessor.GetName(),
+			}
+			om.Annotations = crtbCreatorOwnerAnnotations
+
+			logrus.Infof("[%v] Creating creator clusterRoleTemplateBinding for user %v for cluster %v", ClusterCreateController, creatorID, metaAccessor.GetName())
+			if _, err := l.mgr.Management.ClusterRoleTemplateBindings(metaAccessor.GetName()).Create(&apisv3.ClusterRoleTemplateBinding{
+				ObjectMeta:       om,
+				ClusterName:      metaAccessor.GetName(),
+				RoleTemplateName: role,
+				UserName:         creatorID,
+			}); err != nil && !apierrors.IsAlreadyExists(err) {
+				return obj, err
+			}
+			createdRoles = append(createdRoles, role)
+		}
+
+		roleMap["created"] = createdRoles
+		d, err := json.Marshal(roleMap)
+		if err != nil {
+			return obj, err
+		}
+
+		updateCondition := reflect.DeepEqual(roleMap["required"], createdRoles)
+
+		err = l.updateClusterAnnotationandCondition(cluster, string(d), updateCondition)
+		if err != nil {
+			return obj, err
+		}
+		return obj, nil
+	})
+}
+
+func (l *clusterLifecycle) updateClusterAnnotationandCondition(cluster *apisv3.Cluster, anno string, updateCondition bool) error {
+	sleep := 100
```

Comment:
This looks like it could be refactored to use the retry package? https://pkg.go.dev/k8s.io/client-go/util/retry#OnError

Reply:
Done. The default retry backoff is 5 iterations as opposed to the 4 specified here. But I think that's fine.